### ISSUE TITLE
API additions that allow folks to save updates using json

### ISSUE THREAD

**Issue by ramfox** [qri-io/qri#279]
Let's add:
`/save/meta/peername/datasetname`
`/save/structure/peername/datasetname`
`/save/visConfig/peername/datasetname`
`/save/transform/peername/datasetname`
`/save/data/peername/datasetname`

that allows folks to update the different dataset structures individually, using json.

For example:

`curl -X POST -H "Content-Type: application/json;" http://localhost:2503/save/meta/me/dataset -d '{ "title" : "oh wow so cool" }'`

**Comment by ramfox** [comment]
So realized that if we do it this way:

`curl -X POST -H "Content-Type: application/json;" http://localhost:2503/save/meta/me/dataset -d '{ "title" : "oh wow so cool" }'`

Then the body of the response is the meta.

In that case, there is no way for the user to add a title/message to the save.

If feels important that the user has the opportunity to add a save message if they want, but then it would ruin our cool `{"title":"oh wow how easy"}` json body and turn it into `{"meta":"{"title":"this is less fun"}}` or `{"meta":"{"title":"this is less fun"}, "message":"this is the commit message"}`

Thinking maybe we can allow both?

```
type saveMetaParams struct {
    title string
    message string
    meta []bytes
}

...
// then in saveMetaHandler: 
// json decode response body into saveMetaParams
// if meta is empty: then body was in the form {"title":"updated title"}
// rather then {"meta":{"title":"updated title"}}
// so instead json decode into a slice of bytes
```
Does this make sense?

**Comment by b5** [comment]
hrm, based on that, I think it makes sense to change course. and instead focus on allowing `/save` to accept `Content-Type: application/json;`.

This way, users can supply whichever components they'd like.

#### Metadata only update
```json
{ 
"meta" : { "title" : "oh hai there" }
}
```
#### Metadata with commit tile
```json
{ 
  "commit" : {"title": "bonjour"},
  "meta" : { "title" : "oh hai there" }
}
```

### Metadata and structure, no transform
```json
{
  "meta" : { "title" : "oh hai there" },
  "structure" : { "format" : "json", "schema" : { ... } }
}
```
maor flexible, fewer endpoints to test / be broken

**Comment by ramfox** [comment]
K

**Comment by b5** [comment]
where we at on this? I seem to remember us shipping a bunch of API updates in 0.2.0?

**Comment by ramfox** [comment]
Oh yea, this should have been closed with a PR a week or two ago. Must have forgotten to add it to the close list! My bad.

**Comment by b5** [comment]
lol no worries at all! looks like this was closed by qri-io/qri#280
